From 42147099e0c628882cab36127bd09d371ffd74ab Mon Sep 17 00:00:00 2001
From: Po-Hsu Lin <po-hsu.lin@canonical.com>
Date: Thu, 16 Mar 2023 16:10:48 +0000
Subject: [PATCH] xfstests: add minimal support for zfs

Patch created based on Colin Ian King's commit bb18b1b9 in autotest-client-tests.
Reworked to make it comply with the lastest xfstest-dev commit b90914df.

Currently the following tests work:

001 002 005 006 007 010 011 013 014 023
028 029 035 037 053 067 069 070 075 079
080 084 087 088 089 100 105 112 117 123
124 126 128 129 131 132 141 184 169 192
193 215 221 236 237 241 245 246 247 248
249 257 258 285 286 294 306 307 308 309
310 313 314 317 319

..the following fail with errors (and need
investigating further):

003     access time has changed after accessing file
020     set/get/remove really long names , attr_set failure
062     compare before and after backups - mismatches?
113     aio-stress.4 : -t 20 -s 10m -O -S ???
120     file access time updated on executing file (unexpected)
277     chattr Operation not supported while setting flag
318     Malformed access ACL

..and the rest don't run because of required functionality
not yet implemented in ZFS or in the zfs specific xfstest
common/* scripts:

004     linkat with AT_EMPTY_PATH not supported
008     FALLOC_FL_ZERO_RANGE not supported in fallocate
009     FALLOC_FL_ZERO_RANGE not supported in fallocate
012     FALLOC_FL_ZERO_RANGE not supported in fallocate
015     _scratch_mkfs_sized needs implementing (no-op at the moment)
016     FALLOC_FL_ZERO_RANGE not supported in fallocate
017     fiemap support no implemented in zfs(?)
018     no defrag support in zfs
019     CONFIG_FAIL_MAKE_REQUEST required
021     FALLOC_FL_ZERO_RANGE not supported in fallocate
022     FALLOC_FL_ZERO_RANGE not supported in fallocate
024`    renameat2() - RENAME_NOREPLACE not supported in zfs
025`    renameat2() - RENAME_EXCHANGE not supported in zfs
026     max acls - REDO!!!!
027     _scratch_mkfs_sized needs implementing (no-op at the moment)
030     xfs-io does not support mremap on vivid (why?)
031     FALLOC_FL_COLLAPSE_RANGE no supported by fallocate on zfs yet
                (appeared in 3.15)
032     FALLOC_FL_ZERO_RANGE not supported in fallocate
033     FALLOC_FL_ZERO_RANGE not supported in fallocate
034     needs to be a block device, which it is not (pool device)
036     no O_DIRECT mode on zfs
038     FALLOC_FL_ZERO_RANGE not supported in fallocate
039     needs to be a block device, which it is not (pool device)
040     needs to be a block device, which it is not (pool device)
041     needs to be a block device, which it is not (pool device)
042     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
043     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
044     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
045     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
046     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
047     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
048     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
049     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
050     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
051     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
052     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
054     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
055     ignore for zfs, no zfs equivalent of XFS_IOC_GOINGDOWN
056     needs to be a block device, which it is not (pool device)
057     needs to be a block device, which it is not (pool device)
058     FALLOC_FL_ZERO_RANGE not supported in fallocate
059     needs to be a block device, which it is not (pool device)
060     FALLOC_FL_ZERO_RANGE not supported in fallocate
061     FALLOC_FL_ZERO_RANGE not supported in fallocate
063     FALLOC_FL_ZERO_RANGE not supported in fallocate
064     fiemap support no implemented in zfs(?)
065     needs to be a block device, which it is not (pool device)
066     needs to be a block device, which it is not (pool device)
068     ZFS does not support file system freezing
071     FALLOC_FL_ZERO_RANGE not supported in fallocate
072     FALLOC_FL_COLLAPSE_RANGE no supported by fallocate on zfs yet
                (appeared in 3.15)
073     needs to be a block device, which it is not (pool device)
076     needs to be a block device, which it is not (pool device)
077     _scratch_mkfs_sized needs implementing (no-op at the moment)
078     RENAME_WHITEOUT not supported in ZFS (zfs + overlayfs fixme?)
081     needs to be a block device, which it is not (pool device)
082     fixme - try enable zfs to quota in common/quota and see if this works
083     _scratch_mkfs_sized needs implementing (no-op at the moment)
085     needs to be a block device, which it is not (pool device)
086     FALLOC_FL_ZERO_RANGE not supported in fallocate
091     no O_DIRECT mode on zfs
092     FALLOC_FL_ZERO_RANGE not supported in fallocate
093     IRIX only
097     IRIX only
099     IRIX only
125     no O_DIRECT mode on zfs
130     no O_DIRECT mode on zfs
133     no O_DIRECT mode on zfs
135     no O_DIRECT mode on zfs
198     no O_DIRECT mode on zfs
204     _scratch_mkfs_sized needs implementing (no-op at the moment)
207     no O_DIRECT mode on zfs
208     no O_DIRECT mode on zfs
209     no O_DIRECT mode on zfs
210     no O_DIRECT mode on zfs
211     no O_DIRECT mode on zfs
212     no O_DIRECT mode on zfs
213     FALLOC_FL_ZERO_RANGE not supported in fallocate
214     FALLOC_FL_ZERO_RANGE not supported in fallocate
219     fixme - try enable zfs to quota in common/quota and see if this works
223     FALLOC_FL_ZERO_RANGE not supported in fallocate
224     _scratch_mkfs_sized needs implementing (no-op at the moment)
225     fiemap support no implemented in zfs(?)
226     no O_DIRECT mode on zfs
228     FALLOC_FL_ZERO_RANGE not supported in fallocate
230     fixme - try enable zfs to quota in common/quota and see if this works
231     fixme - try enable zfs to quota in common/quota and see if this works
232     fixme - try enable zfs to quota in common/quota and see if this works
233     fixme - try enable zfs to quota in common/quota and see if this works
234     fixme - try enable zfs to quota in common/quota and see if this works
235     fixme - try enable zfs to quota in common/quota and see if this works
239     no O_DIRECT mode on zfs
240     no O_DIRECT mode on zfs
251     FITRIM not supported on ZFS (xfs specific)
255     FALLOC_FL_ZERO_RANGE not supported in fallocate
256     _scratch_mkfs_sized needs implementing (no-op at the moment)
260     FITRIM not supported on ZFS (xfs specific)
263     no O_DIRECT mode on zfs
269     _scratch_mkfs_sized needs implementing (no-op at the moment)
270     fixme - try enable zfs to quota in common/quota and see if this works
273     _scratch_mkfs_sized needs implementing (no-op at the moment)
274     FALLOC_FL_ZERO_RANGE not supported in fallocate
275     _scratch_mkfs_sized needs implementing (no-op at the moment)
277     chattr failure?
280     fixme - try enable zfs to quota in common/quota and see if this works
288     FITRIM not supported on ZFS (xfs specific)
299     no O_DIRECT mode on zfs
300     no O_DIRECT mode on zfs
311     needs to be a block device, which it is not (pool device)
312     FALLOC_FL_ZERO_RANGE not supported in fallocate
315     FALLOC_FL_ZERO_RANGE not supported in fallocate
316     fiemap support no implemented in zfs(?)
318     duplicate acl?  ..maybe we need to nuke all acls before re-run?
320     _scratch_mkfs_sized needs implementing (no-op at the moment)
321     needs to be a block device, which it is not (pool device)
322     needs to be a block device, which it is not (pool device)
323     no O_DIRECT mode on zfs
324     no defrag support in zfs
325     needs to be a block device, which it is not (pool device)

Signed-off-by: Po-Hsu Lin <po-hsu.lin@canonical.com>
---
 check         |  3 ++-
 common/attr   |  3 +++
 common/config |  4 ++++
 common/rc     | 22 ++++++++++++++++++++++
 4 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/check b/check
index 0bf5b22e..e9979475 100755
--- a/check
+++ b/check
@@ -66,6 +66,7 @@ check options
     -pvfs2		test PVFS2
     -tmpfs		test TMPFS
     -ubifs		test ubifs
+    -zfs		test zfs
     -l			line mode diff
     -udiff		show unified diff (default)
     -n			show me, do not run tests
@@ -280,7 +281,7 @@ while [ $# -gt 0 ]; do
 	case "$1" in
 	-\? | -h | --help) usage ;;
 
-	-nfs|-glusterfs|-cifs|-9p|-fuse|-virtiofs|-pvfs2|-tmpfs|-ubifs)
+	-nfs|-glusterfs|-cifs|-9p|-fuse|-virtiofs|-pvfs2|-tmpfs|-ubifs|-zfs)
 		FSTYP="${1:1}"
 		;;
 	-overlay)
diff --git a/common/attr b/common/attr
index cce4d1b2..b89d17b7 100644
--- a/common/attr
+++ b/common/attr
@@ -25,6 +25,9 @@ _acl_get_max()
 	jfs)
 		echo 8191
 		;;
+	zfs)
+		echo 1024
+		;;
 	f2fs)
 		# If noinline_xattr is enabled, max xattr size should be:
 		# (4096 - 24) - (24 + 4) = 4044
diff --git a/common/config b/common/config
index 6c8cb3a5..a987d4a3 100644
--- a/common/config
+++ b/common/config
@@ -391,6 +391,8 @@ _common_mount_opts()
 	ubifs)
 		echo $UBIFS_MOUNT_OPTIONS
 		;;
+	zfs)
+		;;
 	*)
 		;;
 	esac
@@ -588,6 +590,8 @@ _check_device()
 	fi
 
 	case "$FSTYP" in
+	zfs)
+ 		;;
 	9p|fuse|tmpfs|virtiofs)
 		# 9p, fuse and virtiofs mount tags are just plain strings, so anything is allowed
 		# tmpfs doesn't use mount source, ignore
diff --git a/common/rc b/common/rc
index 6852af79..600b0617 100644
--- a/common/rc
+++ b/common/rc
@@ -602,6 +602,9 @@ _test_mkfs()
     ext2|ext3|ext4)
 	$MKFS_PROG -t $FSTYP -- -F $MKFS_OPTIONS $* $TEST_DEV
 	;;
+    zfs)
+	# do nothing for zfs
+	;;
     *)
 	yes | $MKFS_PROG -t $FSTYP -- $MKFS_OPTIONS $* $TEST_DEV
 	;;
@@ -638,6 +641,9 @@ _try_mkfs_dev()
     ext2|ext3|ext4)
 	$MKFS_PROG -t $FSTYP -- -F $MKFS_OPTIONS $*
 	;;
+    zfs)
+		# do nothing for zfs
+	;;
     xfs)
 	$MKFS_PROG -t $FSTYP -- -f $MKFS_OPTIONS $*
 	;;
@@ -741,6 +747,10 @@ _scratch_mkfs()
 		mkfs_cmd="yes | $MKFS_PROG -t $FSTYP --"
 		mkfs_filter="grep -v -e ^mkfs\.ocfs2"
 		;;
+	zfs)
+		_scratch_cleanup_files
+		return 0
+		;;
 	*)
 		mkfs_cmd="yes | $MKFS_PROG -t $FSTYP --"
 		mkfs_filter="cat"
@@ -1034,6 +1044,8 @@ _scratch_mkfs_sized()
 	bcachefs)
 		$MKFS_PROG -t $FSTYP -- $MKFS_OPTIONS --fs_size=$fssize --block_size=$blocksize $SCRATCH_DEV
 		;;
+	zfs)
+		;;
 	*)
 		_notrun "Filesystem $FSTYP not supported in _scratch_mkfs_sized"
 		;;
@@ -1663,6 +1675,8 @@ _require_scratch_nocheck()
 			_notrun "this test requires a valid \$SCRATCH_MNT"
 		fi
 		;;
+	zfs)
+		;;
 	*)
 		 if [ -z "$SCRATCH_DEV" -o "`_is_block_dev "$SCRATCH_DEV"`" = "" ]
 		 then
@@ -1847,6 +1861,11 @@ _require_test()
 			_notrun "this test requires a valid \$TEST_DIR"
 		fi
 		;;
+	zfs)
+		if [ -z "$TEST_DEV" -o ! -d "$TEST_DIR" ]; then
+			_notrun "this test requires a valid \$TEST_DIR and unique $TEST_DEV"
+		fi
+		;;
 	tmpfs)
 		if [ -z "$TEST_DEV" -o ! -d "$TEST_DIR" ];
 		then
@@ -3171,6 +3190,9 @@ _check_test_fs()
     tmpfs)
 	# no way to check consistency for tmpfs
 	;;
+    zfs)
+	# no way to check consistency for zfs yet
+	;;
     ubifs)
 	# there is no fsck program for ubifs yet
 	;;
-- 
2.17.1

